File: frontend/src/types/indicator.ts
// KDJ Indicator Types

export type KdjSignal = 'golden_cross' | 'death_cross' | 'overbought' | 'oversold' | 'none'

export interface KdjBar {
  open_time: number
  k: number
  d: number
  j: number
  signal: KdjSignal
}

export interface KdjParams {
  n: number
  m1: number
  m2: number
}

export interface KdjResponse {
  data: KdjBar[]
  params: KdjParams
  symbol: string
  interval: string
}

export interface KdjQueryParams {
  symbol: string
  interval: string
  start_time?: number
  end_time?: number
  n?: number
  m1?: number
  m2?: number
}

// ─── MA (Moving Average) ─────────────────────────────────────────────────────

export interface MaBar {
  open_time: number
  ma: number
}

export interface MaResponse {
  data: MaBar[]
  period: number
  symbol: string
  interval: string
}

export interface MaQueryParams {
  symbol: string
  interval: string
  start_time?: number
  end_time?: number
  period?: number
}

// ─── MACD ─────────────────────────────────────────────────────────────────────

export interface MacdBar {
  open_time: number
  macd: number
  signal: number
  histogram: number
}

export interface MacdParams {
  fast_period: number
  slow_period: number
  signal_period: number
}

export interface MacdResponse {
  data: MacdBar[]
  params: MacdParams
  symbol: string
  interval: string
}

export interface MacdQueryParams {
  symbol: string
  interval: string
  start_time?: number
  end_time?: number
  fast_period?: number
  slow_period?: number
  signal_period?: number
}

// ─── RSI ─────────────────────────────────────────────────────────────────────

export interface RsiBar {
  open_time: number
  rsi: number
}

export interface RsiResponse {
  data: RsiBar[]
  period: number
  symbol: string
  interval: string
}

export interface RsiQueryParams {
  symbol: string
  interval: string
  start_time?: number
  end_time?: number
  period?: number
}

// ─── Bollinger Bands ────────────────────────────────────────────────────

export interface BollingerBar {
  open_time: number
  upper: number
  middle: number
  lower: number
}

export interface BollingerParams {
  period: number
  std_dev: number
}

export interface BollingerResponse {
  data: BollingerBar[]
  params: BollingerParams
  symbol: string
  interval: string
}

export interface BollingerQueryParams {
  symbol: string
  interval: string
  start_time?: number
  end_time?: number
  period?: number
  std_dev?: number
}

// ─── EMA (Exponential Moving Average) ────────────────────────────────

export interface EmaBar {
  open_time: number
  ema: number
}

export interface EmaResponse {
  data: EmaBar[]
  period: number
  symbol: string
  interval: string
}

export interface EmaQueryParams {
  symbol: string
  interval: string
  start_time?: number
  end_time?: number
  period?: number
}

// ─── ATR (Average True Range) ────────────────────────────────────────

export interface AtrBar {
  open_time: number
  atr: number
}

export interface AtrResponse {
  data: AtrBar[]
  period: number
  symbol: string
  interval: string
}

export interface AtrQueryParams {
  symbol: string
  interval: string
  start_time?: number
  end_time?: number
  period?: number
}

// ─── Stochastic ─────────────────────────────────────────────────────

export interface StochasticBar {
  open_time: number
  k: number
  d: number
}

export interface StochasticResponse {
  data: StochasticBar[]
  params: { k_period: number; d_period: number; smooth_k: number }
  symbol: string
  interval: string
}

export interface StochasticQueryParams {
  symbol: string
  interval: string
  start_time?: number
  end_time?: number
  k_period?: number
  d_period?: number
  smooth_k?: number
}

File: backend/src/models/schemas/indicator.rs
//! 技术指标 DTO（KDJ/MA/MACD/RSI/Bollinger/EMA/ATR/Stochastic/OBV/Pivot/Fib/Hurst）

use serde::{Deserialize, Serialize};

// ============ KDJ Indicator ============

/// KDJ signal type
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum KdjSignal {
    GoldenCross,
    DeathCross,
    Overbought,
    Oversold,
    #[default]
    None,
}

/// Single KDJ bar result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KdjBar {
    pub open_time: i64,
    pub k: f64,
    pub d: f64,
    pub j: f64,
    pub signal: KdjSignal,
}

/// KDJ calculation parameters
#[derive(Debug, Clone, Serialize)]
pub struct KdjParams {
    pub n: usize,
    pub m1: usize,
    pub m2: usize,
}

/// KDJ API response
#[derive(Debug, Clone, Serialize)]
pub struct KdjResponse {
    pub data: Vec<KdjBar>,
    pub params: KdjParams,
    pub symbol: String,
    pub interval: String,
}

/// KDJ query parameters (incoming from HTTP)
#[derive(Debug, Deserialize)]
pub struct KdjQueryParams {
    pub symbol: Option<String>,
    pub interval: Option<String>,
    pub start_time: Option<i64>,
    pub end_time: Option<i64>,
    pub n: Option<usize>,
    pub m1: Option<usize>,
    pub m2: Option<usize>,
}

// ─── MA (Moving Average) ────────────────────────────────────────────────────

/// Single MA line result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MaBar {
    pub open_time: i64,
    pub ma: f64,
}

/// MA API response
#[derive(Debug, Clone, Serialize)]
pub struct MaResponse {
    pub data: Vec<MaBar>,
    pub period: usize,
    pub symbol: String,
    pub interval: String,
}

/// MA query parameters
#[derive(Debug, Deserialize)]
pub struct MaQueryParams {
    pub symbol: Option<String>,
    pub interval: Option<String>,
    pub start_time: Option<i64>,
    pub end_time: Option<i64>,
    pub period: Option<usize>,
}

// ─── MACD ─────────────────────────────────────────────────────────────────────

/// Single MACD bar result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MacdBar {
    pub open_time: i64,
    pub macd: f64,
    pub signal: f64,
    pub histogram: f64,
}

/// MACD API response
#[derive(Debug, Clone, Serialize)]
pub struct MacdResponse {
    pub data: Vec<MacdBar>,
    pub params: MacdParams,
    pub symbol: String,
    pub interval: String,
}

/// MACD calculation parameters
#[derive(Debug, Clone, Serialize)]
pub struct MacdParams {
    pub fast_period: usize,
    pub slow_period: usize,
    pub signal_period: usize,
}

/// MACD query parameters
#[derive(Debug, Deserialize)]
pub struct MacdQueryParams {
    pub symbol: Option<String>,
    pub interval: Option<String>,
    pub start_time: Option<i64>,
    pub end_time: Option<i64>,
    pub fast_period: Option<usize>,
    pub slow_period: Option<usize>,
    pub signal_period: Option<usize>,
}

// ─── RSI ─────────────────────────────────────────────────────────────────────

/// Single RSI bar result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RsiBar {
    pub open_time: i64,
    pub rsi: f64,
}

/// RSI API response
#[derive(Debug, Clone, Serialize)]
pub struct RsiResponse {
    pub data: Vec<RsiBar>,
    pub period: usize,
    pub symbol: String,
    pub interval: String,
}

/// RSI query parameters
#[derive(Debug, Deserialize)]
pub struct RsiQueryParams {
    pub symbol: Option<String>,
    pub interval: Option<String>,
    pub start_time: Option<i64>,
    pub end_time: Option<i64>,
    pub period: Option<usize>,
}

// ─── Bollinger Bands ────────────────────────────────────────────────────────

/// Single Bollinger Bands bar result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BollingerBar {
    pub open_time: i64,
    pub upper: f64,
    pub middle: f64,
    pub lower: f64,
}

/// Bollinger Bands API response
#[derive(Debug, Clone, Serialize)]
pub struct BollingerResponse {
    pub data: Vec<BollingerBar>,
    pub params: BollingerParams,
    pub symbol: String,
    pub interval: String,
}

/// Bollinger Bands parameters
#[derive(Debug, Clone, Serialize)]
pub struct BollingerParams {
    pub period: usize,
    pub std_dev: f64,
}

/// Bollinger Bands query parameters
#[derive(Debug, Deserialize)]
pub struct BollingerQueryParams {
    pub symbol: Option<String>,
    pub interval: Option<String>,
    pub start_time: Option<i64>,
    pub end_time: Option<i64>,
    pub period: Option<usize>,
    pub std_dev: Option<f64>,
}

/// EMA bar response
#[derive(Debug, Clone, Serialize)]
pub struct EmaResponse {
    pub data: Vec<MaBar>,
    pub period: usize,
    pub symbol: String,
    pub interval: String,
}

/// EMA query parameters
#[derive(Debug, Deserialize)]
pub struct EmaQueryParams {
    pub symbol: Option<String>,
    pub interval: Option<String>,
    pub start_time: Option<i64>,
    pub end_time: Option<i64>,
    pub period: Option<usize>,
}

/// ATR bar response
#[derive(Debug, Clone, Serialize)]
pub struct AtrResponse {
    pub data: Vec<RsiBar>,
    pub period: usize,
    pub symbol: String,
    pub interval: String,
}

/// ATR query parameters
#[derive(Debug, Deserialize)]
pub struct AtrQueryParams {
    pub symbol: Option<String>,
    pub interval: Option<String>,
    pub start_time: Option<i64>,
    pub end_time: Option<i64>,
    pub period: Option<usize>,
}

/// Stochastic bar
#[derive(Debug, Clone, Serialize)]
pub struct StochasticBar {
    pub open_time: i64,
    pub k: f64,
    pub d: f64,
}

/// Stochastic params
#[derive(Debug, Clone, Serialize)]
pub struct StochasticParams {
    pub k_period: usize,
    pub d_period: usize,
    pub smooth_k: usize,
}

/// Stochastic API response
#[derive(Debug, Clone, Serialize)]
pub struct StochasticResponse {
    pub data: Vec<StochasticBar>,
    pub params: StochasticParams,
    pub symbol: String,
    pub interval: String,
}

/// Stochastic query parameters
#[derive(Debug, Deserialize)]
pub struct StochasticQueryParams {
    pub symbol: Option<String>,
    pub interval: Option<String>,
    pub start_time: Option<i64>,
    pub end_time: Option<i64>,
    pub k_period: Option<usize>,
    pub d_period: Option<usize>,
    pub smooth_k: Option<usize>,
}

// ─── OBV (On-Balance Volume) ──────────────────────────────────────────────

/// Single OBV bar result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ObvBar {
    pub open_time: i64,
    pub obv: f64,
}

/// OBV API response
#[derive(Debug, Clone, Serialize)]
pub struct ObvResponse {
    pub data: Vec<ObvBar>,
    pub symbol: String,
    pub interval: String,
}

/// OBV query parameters
#[derive(Debug, Deserialize)]
pub struct ObvQueryParams {
    pub symbol: Option<String>,
    pub interval: Option<String>,
    pub start_time: Option<i64>,
    pub end_time: Option<i64>,
}

// ─── Pivot Points ───────────────────────────────────────────────────────

/// Floor / classical pivot levels (P, R1, S1, R2, S2)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PivotLevels {
    pub p: f64,
    pub r1: f64,
    pub s1: f64,
    pub r2: f64,
    pub s2: f64,
}

/// Single pivot bar
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PivotBar {
    pub open_time: i64,
    pub pivot: PivotLevels,
}

/// Pivot Points API response
#[derive(Debug, Clone, Serialize)]
pub struct PivotResponse {
    pub data: Vec<PivotBar>,
    pub symbol: String,
    pub interval: String,
}

/// Pivot query parameters
#[derive(Debug, Deserialize)]
pub struct PivotQueryParams {
    pub symbol: Option<String>,
    pub interval: Option<String>,
    pub start_time: Option<i64>,
    pub end_time: Option<i64>,
}

// ─── Fibonacci Retracement ─────────────────────────────────────────────

/// Single Fibonacci price level
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FibLevel {
    pub label: String,
    pub ratio: f64,
    pub price: f64,
}

/// Fibonacci retracement response
#[derive(Debug, Clone, Serialize)]
pub struct FibResponse {
    pub high: f64,
    pub low: f64,
    pub range: f64,
    pub levels: Vec<FibLevel>,
}

/// Fibonacci query parameters
#[derive(Debug, Deserialize)]
pub struct FibQueryParams {
    pub high: f64,
    pub low: f64,
}

// ─── Hurst Exponent ────────────────────────────────────────────────────

/// Hurst exponent result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HurstBar {
    pub h: f64,
    pub sample_size: u64,
    pub window_sizes: u32,
}

/// Hurst API response
#[derive(Debug, Clone, Serialize)]
pub struct HurstResponse {
    pub h: f64,
    pub sample_size: u64,
    pub window_sizes: u32,
    pub symbol: String,
    pub interval: String,
    /// `h > 0.5` trending, `h < 0.5` mean-reverting, `h ≈ 0.5` random walk.
    pub interpretation: String,
}

/// Hurst query parameters
#[derive(Debug, Deserialize)]
pub struct HurstQueryParams {
    pub symbol: Option<String>,
    pub interval: Option<String>,
    pub start_time: Option<i64>,
    pub end_time: Option<i64>,
}
